### ISSUE TITLE
Apple M1 ARM support

### ISSUE THREAD

**Issue by luminacious** [go-delve/delve#2246]
Hi All

Hope you're all safe and well.

You're probably already aware of this but just in case.

When trying to debug on Apple M1, it comes back with "rosetta error: failed to allocate vm space for aot".

Just wondering if there's any recommendations please.

Love your work 😊👍

**Comment by aarzilli** [comment]
Go doesn't officially support darwin/arm64 until go1.16, delve won't ufficially support go1.16 until 1.6.0, we don't have access to darwin/arm64 machines to port delve to them so support for delve on darwin/arm64 is unplanned. It's highly unlikely that debuggers can work through rosetta.

**Comment by oxisto** [comment]
I have started with a darwin-arm64 branch on my fork here. https://github.com/oxisto/delve/tree/darwin-arm64 

I basically just started this a couple of hours ago. The native code seems to compile (I am pretty sure it is not functional yet), but Go still throws some link errors. I am not too familiar with that kind of low level programming, so I need some more prep time to fully understand all of it.

Beware, you need a recent Go version that can compile darwin/arm64, i.e. installed using go tip.

**Comment by aarzilli** [comment]
The native code has always had problems, the backend that's being used by default is in pkg/proc/gdbserial, I would focus on that.

**Comment by oxisto** [comment]
> The native code has always had problems, the backend that's being used by default is in pkg/proc/gdbserial, I would focus on that.

Thanks for the pointer in the right direction. I have begun working in this direction on a second fork (https://github.com/oxisto/delve/tree/darwin-arm64-lldb) just by duplicating code for the gdbserver for arm64 now. Spawning an executable works but setting a breakpoint already crashes. I will continue working in this direction then. I will also apply for the mailing list, as you would like to move the discussion there and not on the GitHub issues if I have seen the guidelines correctly.

**Comment by eoarchean** [comment]
Is is possible to test a fork of version currently, or is still still pending other changes? (:

**Comment by andig** [comment]
I'm still seeing these errors on Go 1.16:

    API server listening at: 127.0.0.1:3694
    rosetta error: failed to allocate vm space for aot
    Failed to continue - protocol error E74 during registers read for packet $g;thread:88fee;
    Last known immediate stacktrace (goroutine id 0):
        ?:-1
            <unknown>
        :0
            <unknown>

Should I open a go ticket?

**Comment by s0xzwasd** [comment]
@andig pay attention that you `go version` should be 1.16 darwin/**arm64** instead of 1.16 darwin/amd64.

**Comment by dragonly** [comment]
I installed go with `go version go1.16 darwin/arm64`, and delve 1.6.0, and setup to debug TiDB. which starts fine, but prints wrong var values, and the `n` command behave like drunk 😦

**Comment by aronchick** [comment]
I'm also still seeing this --

```
go version 
go version go1.16.3 darwin/amd64
```

```
❯ dlv version
Delve Debugger
Version: 1.6.0
Build: $Id: 8cc9751909843dd55a46e8ea2a561544f70db34d $
```

Any suggestions? Inside of VSCode, still not working.
```

**Comment by dragonly** [comment]
@aronchick You need to install the arm64 version of golang. You now have `amd64` version of golang, which is not native on Apple M1.

**Comment by aronchick** [comment]
Thanks! That got it.

**Comment by SVilgelm** [comment]
I still have this issue:

I'm using vscode and go and dlv installed using brew:

```shell
21-04-27 16:01 % brew install dlv
Updating Homebrew...
==> Auto-updated Homebrew!
Updated 1 tap (homebrew/core).
==> Updated Formulae
Updated 1 formula.

==> Downloading https://ghcr.io/v2/homebrew/core/delve/manifests/1.6.0
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/delve/blobs/sha256:396965ab785d39bc07d52eb6004f4113d77aa4513d34573e613f03b87567f260
==> Downloading from https://pkg-containers-az.githubusercontent.com/ghcr1/blobs/sha256:396965ab785d39bc07d52eb6004f4113d77aa4513d34573e613f03b87567f260?se=2021-04-27T21%3A10%3A00Z&sig=IDSYg4McG9VCtlqyACF
######################################################################## 100.0%
==> Pouring delve--1.6.0.arm64_big_sur.bottle.tar.gz
🍺  /opt/homebrew/Cellar/delve/1.6.0: 6 files, 13.1MB

21-04-27 16:05 % go version
go version go1.16.3 darwin/arm64

21-04-27 16:08 % dlv version
Delve Debugger
Version: 1.6.0
Build: $Id: 8cc9751909843dd55a46e8ea2a561544f70db34d $
```

but when I run the debugger in vscode I see this error:
```
API server listening at: 127.0.0.1:41753
Failed to continue - protocol error E74 during registers read for packet $g;thread:21ffc;
Last known immediate stacktrace (goroutine id 0):
	?:-1
		<unknown>
	:0
		<unknown>
```


what am I doing wrong? I have go for arm64 and the dlv also is arm64

**Comment by lmx-Hexagram** [comment]
> I still have this issue:
> 
> I'm using vscode and go and dlv installed using brew:
> 
> ```shell
> 21-04-27 16:01 % brew install dlv
> Updating Homebrew...
> ==> Auto-updated Homebrew!
> Updated 1 tap (homebrew/core).
> ==> Updated Formulae
> Updated 1 formula.
> 
> ==> Downloading https://ghcr.io/v2/homebrew/core/delve/manifests/1.6.0
> ######################################################################## 100.0%
> ==> Downloading https://ghcr.io/v2/homebrew/core/delve/blobs/sha256:396965ab785d39bc07d52eb6004f4113d77aa4513d34573e613f03b87567f260
> ==> Downloading from https://pkg-containers-az.githubusercontent.com/ghcr1/blobs/sha256:396965ab785d39bc07d52eb6004f4113d77aa4513d34573e613f03b87567f260?se=2021-04-27T21%3A10%3A00Z&sig=IDSYg4McG9VCtlqyACF
> ######################################################################## 100.0%
> ==> Pouring delve--1.6.0.arm64_big_sur.bottle.tar.gz
> 🍺  /opt/homebrew/Cellar/delve/1.6.0: 6 files, 13.1MB
> 
> 21-04-27 16:05 % go version
> go version go1.16.3 darwin/arm64
> 
> 21-04-27 16:08 % dlv version
> Delve Debugger
> Version: 1.6.0
> Build: $Id: 8cc9751909843dd55a46e8ea2a561544f70db34d $
> 
> % code --version
> 1.55.2
> 3c4e3df9e89829dce27b7b5c24508306b151f30d
> arm64
> ```
> 
> but when I run the debugger in vscode I see this error:
> 
> ```
> API server listening at: 127.0.0.1:41753
> Failed to continue - protocol error E74 during registers read for packet $g;thread:21ffc;
> Last known immediate stacktrace (goroutine id 0):
> 	?:-1
> 		<unknown>
> 	:0
> 		<unknown>
> ```
> 
> what am I doing wrong? I have go for arm64 and the dlv also is arm64 and vscode as well

I meet the same problem after update my m1 chip macOS to 11.3 (20E232)
in this version breakpoint in Goland delve can't get variables and can't move to next point in normal way.
on Vscode I meet the same problem with you.
So I wonder if there is something wrong with new version of macOS

**Comment by Thomascm19** [comment]
> I still have this issue:
> 
> I'm using vscode and go and dlv installed using brew:
> 
> 21-04-27 16:01 % brew install dlv
> Updating Homebrew...
> ==> Auto-updated Homebrew!
> Updated 1 tap (homebrew/core).
> ==> Updated Formulae
> Updated 1 formula.
> 
> ==> Downloading https://ghcr.io/v2/homebrew/core/delve/manifests/1.6.0
> ######################################################################## 100.0%
> ==> Downloading https://ghcr.io/v2/homebrew/core/delve/blobs/sha256:396965ab785d39bc07d52eb6004f4113d77aa4513d34573e613f03b87567f260
> ==> Downloading from https://pkg-containers-az.githubusercontent.com/ghcr1/blobs/sha256:396965ab785d39bc07d52eb6004f4113d77aa4513d34573e613f03b87567f260?se=2021-04-27T21%3A10%3A00Z&sig=IDSYg4McG9VCtlqyACF
> ######################################################################## 100.0%
> ==> Pouring delve--1.6.0.arm64_big_sur.bottle.tar.gz
> 🍺  /opt/homebrew/Cellar/delve/1.6.0: 6 files, 13.1MB
> 
> 21-04-27 16:05 % go version
> go version go1.16.3 darwin/arm64
> 
> 21-04-27 16:08 % dlv version
> Delve Debugger
> Version: 1.6.0
> Build: $Id: 8cc9751909843dd55a46e8ea2a561544f70db34d $
> 
> % code --version
> 1.55.2
> 3c4e3df9e89829dce27b7b5c24508306b151f30d
> arm64
> but when I run the debugger in vscode I see this error:
> 
> API server listening at: 127.0.0.1:41753
> Failed to continue - protocol error E74 during registers read for packet $g;thread:21ffc;
> Last known immediate stacktrace (goroutine id 0):
> 	?:-1
> 		<unknown>
> 	:0
> 		<unknown>
> what am I doing wrong? I have go for arm64 and the dlv also is arm64 and vscode as well


I have the same problem on my mac book air m1 chip, using VScode and GoLand(jetbrains) i'm using macOS to 11.3 (20E232)

**Comment by s0xzwasd** [comment]
@Thomascm19 @lmx-Hexagram @SVilgelm please see the corresponding issue: https://github.com/go-delve/delve/issues/2436

Especially the possible workarounds: https://github.com/go-delve/delve/issues/2436#issuecomment-821857172 & https://github.com/go-delve/delve/issues/2436#issuecomment-828092306

**Comment by SVilgelm** [comment]
@s0xzwasd thank you!

**Comment by risadharma** [comment]
@s0xzwasd thanks for referring us to the corresponding issue. It works for me

**Comment by hielfx** [comment]
In my case, I fixed it by downloading "Apple Silicon" version of vscode, instead of clicking the big download one, and then deleting and reinstalling all the go-tools.
![imagen](https://user-images.githubusercontent.com/7794231/123427507-012e0980-d5c5-11eb-98a7-406522ae80df.png)

**Comment by microyahoo** [comment]
The issue still exists in new version again.  @oxisto 
```
🍺 /Users/zhl/go/src/github.com/grpc/grpc-go/examples/helloworld ☞ git:(master) ✗ dlv exec client
Type 'help' for list of commands.
(dlv) b main.main
Breakpoint 1 (enabled) set at 0x13d919b for main.main() ./greeter_client/main.go:37
(dlv) c
rosetta error: failed to allocate vm space for aot
Stopped at: 0x0
=>   1:	no source available
(dlv) exit
🍺 /Users/zhl/go/src/github.com/grpc/grpc-go/examples/helloworld ☞ git:(master) ✗ dlv version
Delve Debugger
Version: 1.6.1
Build: de117a2f4f16350a59b06970693d383613399782
🍺 /Users/zhl/go/src/github.com/grpc/grpc-go/examples/helloworld ☞ git:(master) ✗ go env
GO111MODULE="on"
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/zhl/Library/Caches/go-build"
GOENV="/Users/zhl/Library/Application Support/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/zhl/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/zhl/go"
GOPRIVATE=""
GOPROXY="https://goproxy.cn"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GOVCS=""
GOVERSION="go1.16.4"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/zhl/go/src/github.com/grpc/grpc-go/examples/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch x86_64 -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/x2/p8kghjr100j209q5zspdwhgc0000gn/T/go-build2395424900=/tmp/go-build -gno-record-gcc-switches -fno-common"
```

**Comment by derekparker** [comment]
> The issue still exists in new version again. @oxisto

You’re running the amd64 tool chain on apple silicon. You should instead download the Darwin/arm64 version which runs natively on Apple silicon.

**Comment by oxisto** [comment]
> The issue still exists in new version again. @oxisto
> 
> ```
> 🍺 /Users/zhl/go/src/github.com/grpc/grpc-go/examples/helloworld ☞ git:(master) ✗ dlv exec client
> Type 'help' for list of commands.
> (dlv) b main.main
> Breakpoint 1 (enabled) set at 0x13d919b for main.main() ./greeter_client/main.go:37
> (dlv) c
> rosetta error: failed to allocate vm space for aot
> Stopped at: 0x0
> =>   1:	no source available
> (dlv) exit
> 🍺 /Users/zhl/go/src/github.com/grpc/grpc-go/examples/helloworld ☞ git:(master) ✗ dlv version
> Delve Debugger
> Version: 1.6.1
> Build: de117a2f4f16350a59b06970693d383613399782
> 🍺 /Users/zhl/go/src/github.com/grpc/grpc-go/examples/helloworld ☞ git:(master) ✗ go env
> GO111MODULE="on"
> GOARCH="amd64"
> GOBIN=""
> GOCACHE="/Users/zhl/Library/Caches/go-build"
> GOENV="/Users/zhl/Library/Application Support/go/env"
> GOEXE=""
> GOFLAGS=""
> GOHOSTARCH="amd64"
> GOHOSTOS="darwin"
> GOINSECURE=""
> GOMODCACHE="/Users/zhl/go/pkg/mod"
> GONOPROXY=""
> GONOSUMDB=""
> GOOS="darwin"
> GOPATH="/Users/zhl/go"
> GOPRIVATE=""
> GOPROXY="https://goproxy.cn"
> GOROOT="/usr/local/go"
> GOSUMDB="sum.golang.org"
> GOTMPDIR=""
> GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
> GOVCS=""
> GOVERSION="go1.16.4"
> GCCGO="gccgo"
> AR="ar"
> CC="clang"
> CXX="clang++"
> CGO_ENABLED="1"
> GOMOD="/Users/zhl/go/src/github.com/grpc/grpc-go/examples/go.mod"
> CGO_CFLAGS="-g -O2"
> CGO_CPPFLAGS=""
> CGO_CXXFLAGS="-g -O2"
> CGO_FFLAGS="-g -O2"
> CGO_LDFLAGS="-g -O2"
> PKG_CONFIG="pkg-config"
> GOGCCFLAGS="-fPIC -arch x86_64 -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/x2/p8kghjr100j209q5zspdwhgc0000gn/T/go-build2395424900=/tmp/go-build -gno-record-gcc-switches -fno-common"
> ```

According to `go env` you are using an amd64 version of Go. You need to install the arm64 version of Go to make `dlv` function properly on an M1 machine.

**Comment by EsmaeilMazahery** [comment]
make sure vs code or goland are arm64 and in vs code reinstall tools

**Comment by Schachte** [comment]
```
extender.InitTcpServer(ctx, reset)
Unable to evaluate expression: function calls not allowed without using 'call'
call extender.InitTcpServer(ctx, reset)
Unable to evaluate expression: backend does not support function calls
```

I'm still unable to invoke function calls using DLV via the debug console.

**Comment by Jaydonchong** [comment]
Chiming in here as I ran into the same `E74 during registers read for packet $g;thread:...` issue while trying to run `go1.16` on vscode. I was able to resolve this by updating delve to use 1.6.1 for the fix in wrong g address (not exactly sure what this is for but I'm guessing it's got to do with reading memory address) ([src](https://github.com/go-delve/delve/blob/master/CHANGELOG.md#:~:text=Correctly%20read%20G%20address%20on%20linux/arm64)). I was also facing an issue whereby my filepath wasn't resolving properly in program and the update solved it as well

Not sure if this is the clean way to do it but the way I updated dlv from 1.6 was :
1. `rm -rf $GOPATH/bin/dlv`
2. run `go install github.com/go-delve/delve/cmd/dlv@v1.6.1`
3. restarted vscode and the debug was working fine

Hope this helps 😄
